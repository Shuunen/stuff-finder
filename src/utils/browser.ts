import { dom, emit, sleep, storage, tw } from 'shuutils'
import { JSON_HEADERS } from '../constants'
import type { AirtableRecord, AppToasterShowEvent } from '../types'
import { urlToUuid } from './url'

export const button = (content: string, classes = '', secondary = false): HTMLButtonElement => {
  const theme = secondary ? tw('opacity-80 hover:opacity-100') : tw('from-purple-500 to-purple-700 text-white hover:from-purple-700 hover:to-purple-900')
  const button_ = dom('button', tw(`h-10 max-w-xs rounded bg-gradient-to-tr px-12 text-lg shadow-md transition-all duration-200 ease-in-out hover:shadow-lg sm:px-6 md:text-base ${theme} ${classes}`), content)
  button_.type = 'button'
  return button_
}

/* eslint-disable no-console */
export const logger = {
  error: (message: string, ...data: readonly unknown[]): void => { console.error(message, ...data) },
  showError: (message: string, ...data: readonly unknown[]): void => {
    console.error(message, ...data)
    emit<AppToasterShowEvent>('app-toaster--show', { type: 'error', message })
  },
  log: (message: string, ...data: readonly unknown[]): void => { console.log(message, ...data) },
  showLog: (message: string, ...data: readonly unknown[]): void => {
    console.log(message, ...data)
    emit<AppToasterShowEvent>('app-toaster--show', { type: 'info', message })
  },
}
/* eslint-enable no-console */

const request = async <T> (method: 'get' | 'patch' | 'post', url: string, data?: Record<string, unknown>): Promise<T> => {
  const options: RequestInit = { headers: JSON_HEADERS, method }
  if (data) options.body = JSON.stringify(data)
  return fetch(url, options).then(async (response: Response) => response.json()).catch(error => { logger.showError((error as Error).message) }) as Promise<T>
}

export const patch = async (url: string, data: Record<string, unknown>): Promise<AirtableRecord> => request('patch', url, data)

export const post = async (url: string, data: Record<string, unknown>): Promise<AirtableRecord> => request('post', url, data)

export const get = async <T> (url: string): Promise<T> => {
  const uuid = urlToUuid(url)
  const cached = storage.get<T>(uuid, undefined, sessionStorage)
  if (cached !== undefined) return cached
  const response = await request<T>('get', url)
  storage.set(uuid, response, sessionStorage)
  return response
}

/* eslint-disable no-restricted-properties, unicorn/prefer-spread */
export const find = {
  one: <T extends Element = Element> (selector: string, context: Document | Element = document): T => {
    const element = context.querySelector<T>(selector)
    if (!element) throw new Error(`no element found for selector "${selector}"`)
    return element
  },
  oneOrNone: <T extends Element = Element> (selector: string, context: Document | Element = document): T | null => context.querySelector<T>(selector),
  all: <T extends Element = Element> (selector: string, context: Document | Element = document): T[] => {
    const elements = context.querySelectorAll<T>(selector)
    if (elements.length === 0) throw new Error(`no elements found for selector "${selector}"`)
    return Array.from(elements)
  },
  allOrNone: <T extends Element = Element> (selector: string, context: Document | Element = document): T[] => Array.from(context.querySelectorAll<T>(selector)),
}
/* eslint-enable no-restricted-properties, unicorn/prefer-spread */

export const fadeIn = async (element: HTMLElement): Promise<void> => {
  if (!element.classList.contains('app-hide')) { logger.error('please add "app-hide" class before mounting dom element and then call fade-in'); return }
  element.classList.remove('hidden')
  await sleep(10)
  element.style.opacity = '1'
}

export const fadeOut = async (element: HTMLElement, destroy = false): Promise<void> => {
  element.classList.add('app-hide')
  element.style.opacity = '0'
  await sleep(350)
  element.classList.add('hidden')
  if (!destroy) return
  await sleep(350)
  element.remove()
}

export const valuesToOptions = (array: readonly string[], selected?: string): string => {
  return array.map(value => `<option value="${value}" ${selected === value ? 'selected' : ''}>${value}</option>`).join('')
}

export const isVisible = (element: Element | HTMLElement | undefined): boolean => {
  if (!element) {
    logger.error('element is undefined, so it is not visible')
    return false
  }
  const { top, bottom, width, height } = element.getBoundingClientRect()
  const visible = width > 0 && height > 0 && top >= 0 && bottom <= window.innerHeight
  // logger.log(`element is ${visible ? '' : 'not '}visible`, element)
  return visible
}
