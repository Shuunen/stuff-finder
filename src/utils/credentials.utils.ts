import type { AppCredentials } from '../types'

export function parseClipboard (clipboard: string) {
  // clipboard should contains something like : "appABC
  // patXYZ.123
  // my-table
  // my-view
  // azerty4567"
  const regex = /"(?<base>app\w+)\n(?<token>pat[\w.]+)\n(?<table>[\w-]+)\n(?<view>[\w-]+)\n(?<wrap>[\w-]+)"/u
  const { base = '', table = '', token = '', view = '', wrap = '' } = regex.exec(clipboard)?.groups ?? {}
  return { base, table, token, view, wrap } satisfies AppCredentials
}
