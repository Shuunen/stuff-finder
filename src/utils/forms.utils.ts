/* eslint-disable jsdoc/require-jsdoc */
import { objectEqual } from 'shuutils'

type FormFieldType = 'checkbox' | 'select' | 'text'
type FormFieldOptions = { label: string; value: string }[]
type FormFieldBase = {
  columns: number
  isRequired: boolean
  isValid: boolean
  isVisible: boolean
  label: string
  link: string
  options?: FormFieldOptions
  regex: RegExp
  unit: string
}
type FormFieldText = FormFieldBase & { type: 'text'; value: string }
type FormFieldSelect = FormFieldBase & { options: FormFieldOptions; type: 'select'; value: string }
type FormFieldCheckbox = FormFieldBase & { type: 'checkbox'; value: boolean } // eslint-disable-line @typescript-eslint/naming-convention
type FormField<Type extends FormFieldType> = Type extends 'text' ? FormFieldText : Type extends 'select' ? FormFieldSelect : FormFieldCheckbox

function fieldRegex (regex?: RegExp, minLength = 3, maxLength = 100) {
  return regex ?? new RegExp(`^[-,\\d\\p{L}\\s/&]{${minLength},${maxLength}}$`, 'u')
}

export type Form = {
  columns?: number
  errorMessage: string
  fields: Record<string, FormField<FormFieldType>>
  isTouched: boolean
  isValid: boolean
}

export function validateForm<FormType extends Form> (form: FormType) {
  let errorMessage = ''
  const updatedFields = Object.entries(form.fields).reduce((accumulator, [field, { isRequired, label, regex, value }]) => { // eslint-disable-line unicorn/no-array-reduce
    const isValid = !isRequired && (typeof value === 'string' && value === '') || typeof value === 'string' && regex.test(value) || typeof value === 'boolean'
    if (!isValid) errorMessage = value === '' ? `${label} is required` : `${label} is invalid, "${String(value)}" should match ${String(regex)}`
    // biome-ignore lint/performance/noAccumulatingSpread: this whole validation is a joke, we should use a proper library
    return { ...accumulator, [field]: { ...form.fields[field], isValid } }
  }, {})
  const isFormValid = errorMessage === ''
  const updatedForm: FormType = { ...form, errorMessage, fields: updatedFields, isTouched: form.isTouched, isValid: isFormValid } satisfies FormType
  const hasChanged = !objectEqual(form, updatedForm, true)
  return { hasChanged, updatedForm }
}

// eslint-disable-next-line complexity
export function createTextField (parameters: Partial<Pick<FormFieldText, 'columns' | 'isRequired' | 'isValid' | 'isVisible' | 'link' | 'regex' | 'unit' | 'value'>> & Pick<FormFieldText, 'label'> & { maxLength?: number; minLength?: number }) {
  const { columns = 1, isRequired = false, isValid = false, isVisible = true, label = '', link = '', maxLength = 100, minLength = 3, regex, unit = '', value = '' } = parameters
  return { columns, isRequired, isValid, isVisible, label, link, regex: fieldRegex(regex, minLength, maxLength), type: 'text', unit, value } satisfies FormFieldText
}

export function createCheckboxField (parameters: Partial<Pick<FormFieldCheckbox, 'columns' | 'isRequired' | 'isValid' | 'isVisible' | 'link' | 'value'>> & Pick<FormFieldCheckbox, 'label'>) {
  const { columns = 1, isRequired = false, isValid = false, isVisible = true, label = '', link = '', value = false } = parameters // eslint-disable-line @typescript-eslint/naming-convention
  return { columns, isRequired, isValid, isVisible, label, link, regex: fieldRegex(), type: 'checkbox', unit: '', value } satisfies FormFieldCheckbox // eslint-disable-line @typescript-eslint/naming-convention
}

export function createSelectField (parameters: Partial<Pick<FormFieldSelect, 'columns' | 'isRequired' | 'isValid' | 'isVisible' | 'link' | 'regex' | 'value'>> & Pick<FormFieldSelect, 'label' | 'options'>) {
  const { columns = 1, isRequired = false, isValid = false, isVisible = true, label = '', link = '', options, regex, value = '' } = parameters
  return { columns, isRequired, isValid, isVisible, label, link, options, regex: fieldRegex(regex), type: 'select', unit: '', value } satisfies FormFieldSelect
}

export function optionsToLabels (values?: FormFieldOptions) {
  return values?.map(({ label }) => label) ?? []
}

export type { FormFieldCheckbox, FormFieldSelect, FormFieldText }


export function alignClipboard (text: string) {
  return text.replace(/: ""(?<thing>[,\n])/gu, ': "__EMPTY__"$<thing>')
    .replace(/""/gu, '"')
    .replace(/"__EMPTY__"/gu, '""')
    .replace('"{', '{')
    .replace('}"', '}') // need to replace double double quotes with single double quotes (Google Sheet issue -.-'''''')
}
