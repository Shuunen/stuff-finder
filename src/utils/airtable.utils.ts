/* eslint-disable functional/immutable-data */
import { safeParse } from 'valibot'
import { emptyItem } from '../constants'
import { del, get, patch, post } from './browser.utils'
import { logger } from './logger.utils'
import { airtableDeleteResponseSchema, airtableMultipleResponseSchema, airtableSingleResponseSchema, type AirtableSingleRecordResponse, type Item, type ItemField, type ItemPhoto } from './parsers.utils'
import { state } from './state.utils'

const airtableBaseUrl = 'https://api.airtable.com/v0'

export const airtableMaxRequestPerSecond = 5

// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: <explanation>
export function getItemFieldsToPush (data: Item, currentState = state) { // eslint-disable-line max-statements, complexity, sonarjs/cognitive-complexity
  const fields: Partial<AirtableSingleRecordResponse['fields']> = {}
  if (data.barcode.length > 0) fields.barcode = data.barcode
  if (data.box.length > 0) fields.box = data.box
  if (data.brand.length > 0) fields.brand = data.brand
  if (data.category.length > 0) fields.category = data.category
  if (data.details.length > 0) fields.details = data.details
  if (data.drawer.length > 0) fields.drawer = data.drawer
  if (data.location.length > 0) fields.location = data.location
  if (data.name.length > 0) fields.name = data.name
  /* c8 ignore next 2 */
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (data.photo !== undefined && data.photo.length > 0 && fields.photo?.[0]?.url !== data.photo[0]?.url) fields.photo = [{ url: data.photo[0]?.url } as unknown as ItemPhoto] // we don't need the whole object
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  if (data.price !== undefined && data.price > -1) fields.price = data.price
  if (data.reference.length > 0) fields.reference = data.reference
  if (data.status.length > 0) fields.status = data.status
  fields['ref-printed'] = data['ref-printed']
  if (data.id.length > 0) {
    const existing = currentState.items.find(existingItem => existingItem.id === data.id)
    if (!existing) throw new Error('existing item not found locally')
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const dataFields = Object.keys(fields) as ItemField[]
    for (const field of dataFields) {
      /* c8 ignore next 2 */
      if (field === 'id') continue // eslint-disable-line no-continue
      const hasSamePhoto = (field === 'photo' && existing.photo && existing.photo[0]?.url === fields.photo?.[0]?.url) ?? false
      const hasSameValue = existing[field] === fields[field]
      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
      if (hasSamePhoto || hasSameValue) delete fields[field]
    }
  }
  return fields
}

export function airtableRecordToItem (record: AirtableSingleRecordResponse) {
  return {
    ...emptyItem,
    ...record.fields,
    id: record.id,
  }
}

export function isLocalAndRemoteSync (records: AirtableSingleRecordResponse[], currentState = state) {
  const record = records.at(0)
  if (!record) throw new Error('remoteFirst is undefined')
  const remoteFirst = airtableRecordToItem(record)
  const localFirst = currentState.items.at(0)
  const isFirst = localFirst === undefined ? false : remoteFirst.id === localFirst.id && remoteFirst['updated-on'] === localFirst['updated-on']
  if (isFirst) return true
  const localLast = currentState.items.at(-1) // eslint-disable-line @typescript-eslint/no-magic-numbers
  return localLast === undefined ? false : remoteFirst.id === localLast.id && remoteFirst['updated-on'] === localLast['updated-on']
}

export async function getOneItem (id: Item['id'], getMethod = get) {
  const { base, table } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  const result = safeParse(airtableSingleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return airtableRecordToItem(result.output)
}

export async function getAllItems (offset?: string, getMethod = get) {
  const offsetParameter = offset === undefined ? '' : `&offset=${offset}`
  const sortByUpdatedFirst = '&sort%5B0%5D%5Bfield%5D=updated-on&sort%5B0%5D%5Bdirection%5D=desc'
  const { base, table, view } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}?view=${view}${offsetParameter}${sortByUpdatedFirst}`
  const result = safeParse(airtableMultipleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return result.output
}

export async function deleteItemRemotely (id: Item['id'], currentState = state, delMethod = del) {
  const { base, table } = currentState.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  return safeParse(airtableDeleteResponseSchema, await delMethod(url))
}

// eslint-disable-next-line @typescript-eslint/max-params
export async function pushItemRemotely (item: Item, currentState = state, postMethod = post, patchMethod = patch) {
  const fields = getItemFieldsToPush(item, currentState)
  if (Object.keys(fields).length === 0) {
    logger.showLog('no update to push')
    return { output: { fields, id: item.id }, success: false } // eslint-disable-line @typescript-eslint/naming-convention
  }
  const data = { fields }
  const { base, table } = currentState.credentials
  const baseUrl = `${airtableBaseUrl}/${base}/${table}`
  if (item.id === '') return safeParse(airtableSingleResponseSchema, await postMethod(baseUrl, data))
  const url = `${baseUrl}/${item.id}`
  return safeParse(airtableSingleResponseSchema, await patchMethod(url, data))
}
