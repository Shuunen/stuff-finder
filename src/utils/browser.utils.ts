/* eslint-disable jsdoc/require-jsdoc */
import { logger } from './logger.utils'

async function request (method: 'DELETE' | 'GET' | 'PATCH' | 'POST', url: string, data?: Record<string, unknown>) {
  const options: RequestInit = { method }
  if (data) options.body = JSON.stringify(data)
  const response = await fetch(url, options).catch((error: unknown) => { logger.showError(error) })
  // eslint-disable-next-line no-restricted-syntax
  if (!response) throw new Error('no response')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return await response.json() as unknown
}

// biome-ignore lint/suspicious/useAwait: <explanation>
export async function get (url: string) {
  return request('GET', url)
}

export function valuesToOptions (array: readonly string[], selected?: string) {
  return array.map(value => `<option value="${value}" ${selected === value ? 'selected' : ''}>${value}</option>`).join('')
}

export function setPageTitle (title: string) {
  document.title = `${title} - Stuff Finder`
}

export function clearElementsForPrint () {
  logger.info('clearing elements for print')
  const selector = [
    '#synology-download-notification-stack',
    'synology-download-content',
    '[at-magnifier-wrapper]',
    '.shu-toast',
  ]
  const elements = document.querySelectorAll(selector.join(','))
  for (const element of elements) element.remove()
}
