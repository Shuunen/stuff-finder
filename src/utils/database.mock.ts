/* eslint-disable @typescript-eslint/consistent-type-assertions */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/max-params */
/* eslint-disable @typescript-eslint/explicit-member-accessibility */
/* eslint-disable jsdoc/require-jsdoc */
/* eslint-disable no-restricted-syntax */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable max-classes-per-file */
import type { Models } from 'appwrite'
import { functionReturningVoid, nbDaysInWeek, sleep } from 'shuutils'
import { vi } from 'vitest'
import type { ItemModel } from '../types/item.types'
import { mockItemModel } from './mock.utils'

export function mockFile (data: Partial<Models.File> = {}) {
  return {
    '$createdAt': '2020-03-01T00:00:00.000Z',
    '$id': 'some-image-file-uuid-xyz',
    '$permissions': [],
    '$updatedAt': '2021-08-01T00:00:00.000Z',
    bucketId: 'bucketA',
    chunksTotal: 1,
    chunksUploaded: 1,
    mimeType: 'image/jpeg',
    name: 'fileName-a.jpg',
    signature: 'signature-a',
    sizeOriginal: 123_456,
    ...data,
  } satisfies Models.File as Models.File
}

const createFile = vi.fn(async (bucketId: string, fileId: string, _file: File) => {
  await sleep(nbDaysInWeek)
  return { $id: fileId, bucketId, isThisMockedDataFromMock: true }
})

const deleteFile = vi.fn(async (bucketId: string, fileId: string, _file: File) => {
  await sleep(nbDaysInWeek)
  return { $id: fileId, bucketId, isThisMockedDataFromMock: true }
})

const listFiles = vi.fn(async (_bucketId: string, _queries: [{ limit: number }, { offset: number }]) => {
  await sleep(nbDaysInWeek)
  return { files: [], total: 0 } satisfies Models.FileList as Models.FileList
})

const createDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string, data: object) => {
  await sleep(nbDaysInWeek)
  const item = mockItemModel({ $collectionId: collectionId, $databaseId: databaseId, $id: documentId, ...data })
  return item satisfies Models.Document as Models.Document
})

const deleteDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string) => {
  await sleep(nbDaysInWeek)
  return { $id: documentId, collectionId, databaseId, isThisMockedDataFromMock: true }
})

const listDocuments = vi.fn(async (_databaseId: string, _collectionId: string) => {
  await sleep(nbDaysInWeek)
  return { documents: [], total: 0 } satisfies Models.DocumentList<ItemModel> as Models.DocumentList<ItemModel>
})

const updateDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string, data: object) => {
  await sleep(nbDaysInWeek)
  const item = mockItemModel({ $collectionId: collectionId, $databaseId: databaseId, $id: documentId, ...data })
  return item satisfies Models.Document as Models.Document
})

class Databases {
  createDocument = createDocument
  deleteDocument = deleteDocument
  listDocuments = listDocuments
  updateDocument = updateDocument
  constructor (client?: Client) {
    if (client) functionReturningVoid()
  }
}

class Client {
  constructor () {
    functionReturningVoid()
  }
  setProject (project: string) {
    if (project) functionReturningVoid()
    return this
  }
}

class Storage {
  createFile = createFile
  deleteFile = deleteFile
  listFiles = listFiles
  constructor (client?: Client) {
    if (client) functionReturningVoid()
  }
}

const Query = {
  limit: vi.fn((nb: number) => ({ isThisMockedDataFromMock: true, limit: nb })),
  offset: vi.fn((nb: number) => ({ isThisMockedDataFromMock: true, offset: nb })),
}

function reset () {
  createDocument.mockClear()
  createFile.mockClear()
  listFiles.mockClear()
  deleteDocument.mockClear()
  deleteFile.mockClear()
  listDocuments.mockClear()
  updateDocument.mockClear()
  Query.limit.mockClear()
  Query.offset.mockClear()
}

export const databaseMock = {
  appwrite: { Client, Databases, Query, Storage },
  createDocument, createFile, deleteDocument, deleteFile, listDocuments, listFiles, Query, reset, updateDocument,
}
