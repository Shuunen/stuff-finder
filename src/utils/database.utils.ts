/* eslint-disable no-await-in-loop */
import { Client, Databases, type Models, Query, Storage } from 'appwrite'
import { Result, dateIso10, storage as lsStorage, nbHueMax, nbPercentMax, nbSpacesIndent, sleep, slugify, toastSuccess } from 'shuutils'
import { safeParse } from 'valibot'
import { defaultImage, uuidMaxLength } from '../constants'
import type { Item, ItemModel } from '../types/item.types'
import { logger } from './logger.utils'
/* eslint-disable jsdoc/require-jsdoc */
import { itemModelSchema, itemSchema, itemsSchema } from './parsers.utils'
import { state } from './state.utils'
import { normalizePhotoUrl } from './url.utils'

const client = new Client()
const database = new Databases(client)
const storage = new Storage(client)
const projectId = 'stuff-finder'
client.setProject(projectId)

export function itemIdToImageUrl (id: Item['$id']) {
  return `https://cloud.appwrite.io/v1/storage/buckets/${state.credentials.bucketId}/files/${id}/view?project=${projectId}&mode=admin`
}

const appWritePhotoIdRegex = /cloud\.appwrite\.io\/v1\/storage\/buckets\/\w+\/files\/(?<id>[^/]+)\/view/u

export function getAppWriteIdFromUrl (url: string) {
  const { id } = (appWritePhotoIdRegex.exec(url))?.groups ?? {}
  return id
}

const urlRegex = /https?:\/\/\S+/u

function isUrl (text: string) {
  return urlRegex.test(text)
}

export function itemPhotoToImageUrl (photo?: string) {
  logger.debug('itemPhotoToImageUrl', { photo })
  if (photo === undefined || photo === '') return defaultImage
  if (!isUrl(photo)) return itemIdToImageUrl(photo)
  const id = getAppWriteIdFromUrl(photo)
  if (id === undefined) return normalizePhotoUrl(photo) // external url
  return itemIdToImageUrl(id) // bucket url
}

export function itemToImageUrl (item?: Item) {
  const photo = item?.photos[0]
  return itemPhotoToImageUrl(photo)
}

export function fileTypeToExtension (type: string) {
  if (type === 'image/jpeg') return Result.ok('jpg')
  if (type === 'image/svg+xml') return Result.ok('svg')
  if (type === 'image/avif') return Result.ok('avif')
  if (type === 'image/webp') return Result.ok('webp')
  return Result.error(`un-wanted file type : ${type}`)
}

export async function deleteImageRemotely (id: string) {
  const result = await Result.trySafe(storage.deleteFile(state.credentials.bucketId, id))
  if (result.ok) logger.success(`image "${id}" deleted successfully`)
  else logger.error(`image "${id}" deletion failed`, result.error)
  return result
}

// eslint-disable-next-line max-statements
export async function uploadImage (fileName: string, url: string) {
  const blob = await fetch(url).then(async response => response.blob())
  const extension = fileTypeToExtension(blob.type)
  if (!extension.ok) return Result.ok(url)
  const hasExtension = fileName.includes('.')
  const finalFileName = hasExtension ? fileName : `${fileName}.${extension.value}`
  const file = new File([blob], finalFileName, { type: blob.type })
  const id = slugify(finalFileName.replace(/[_.]/gu, '-')).slice(0, uuidMaxLength)
  let upload = await Result.trySafe(storage.createFile(state.credentials.bucketId, id, file))
  if (!upload.ok) {
    logger.error('uploadImage failed', upload.error)
    if (String(upload.error).includes('requested ID already exists')) await deleteImageRemotely(id)
    upload = await Result.trySafe(storage.createFile(state.credentials.bucketId, id, file)) // retry
  }
  if (!upload.ok) return Result.ok(url)
  return Result.ok(upload.value.$id)
}

export function downloadBlob (blob: Blob, fileName: string) {
  const link = document.createElement('a')
  link.href = URL.createObjectURL(blob)
  link.download = fileName
  document.body.append(link)
  link.click()
  link.remove()
  URL.revokeObjectURL(link.href)
  toastSuccess('Download started')
}

export function downloadObject (data: Record<string, unknown> | unknown[], fileName: string) {
  const json = JSON.stringify(data, undefined, nbSpacesIndent)
  const blob = new Blob([json], { type: 'application/json' })
  downloadBlob(blob, fileName)
}

export async function downloadUrl (url: string, fileName: string) {
  const result = await fetch(url)
  const blob = await result.blob()
  downloadBlob(blob, fileName)
}

export async function listImages (bucketId = state.credentials.bucketId) {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const images = [] as Models.File[]
  let offset = 0
  let shouldCheckNextPage = true
  while (shouldCheckNextPage) {
    const result = await Result.trySafe(storage.listFiles(bucketId, [Query.limit(nbPercentMax), Query.offset(offset)]))
    if (!result.ok) return result
    if (result.value.files.length === 0) shouldCheckNextPage = false
    else {
      images.push(...result.value.files)
      offset += nbPercentMax
    }
  }
  return Result.ok(images)
}

// eslint-disable-next-line max-statements
export async function downloadImages (bucketId = state.credentials.bucketId) {
  const result = await listImages(bucketId)
  if (!result.ok) return result
  const downloadedImages = lsStorage.get<string[]>('downloadedImages', [])
  /* c8 ignore next 8 */
  for (const file of result.value) {
    if (downloadedImages.includes(file.$id)) continue
    downloadedImages.push(file.$id)
    const url = itemIdToImageUrl(file.$id)
    await downloadUrl(url, file.name)
    await sleep(nbHueMax)
    lsStorage.set('downloadedImages', downloadedImages)
  }
  return Result.ok('images downloaded successfully')
}

// eslint-disable-next-line max-statements
export async function getItemsRemotely () {
  const items: ItemModel[] = []
  let offset = 0
  let shouldCheckNextPage = true
  while (shouldCheckNextPage) {
    const result = await Result.trySafe(database.listDocuments<ItemModel>(state.credentials.databaseId, state.credentials.collectionId, [Query.limit(nbPercentMax), Query.offset(offset)]))
    if (!result.ok) return result
    if (result.value.documents.length === 0) shouldCheckNextPage = false
    else {
      items.push(...result.value.documents)
      offset += nbPercentMax
    }
  }
  state.itemsTimestamp = Date.now()
  const result = safeParse(itemsSchema, items)
  if (!result.success) return Result.error(result.issues.map(issue => issue.message).join(', '))
  return Result.ok(result.output)
}

export async function downloadItems () {
  const result = await getItemsRemotely()
  /* c8 ignore next */
  if (!result.ok) return result
  downloadObject(result.value, `${dateIso10()}_${projectId}_items.json`)
  return Result.ok('items downloaded successfully')
}

export function getItemId (item: Item) {
  const id = slugify(item.reference || item.name).slice(0, uuidMaxLength)
  if (id.length === 0) return Result.error(`item id is empty in ${JSON.stringify(item)}`)
  return Result.ok(id)
}

export function removeAppWriteFields (item: Record<string, unknown>) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { $id, ...rest } = structuredClone(item)
  return rest
}

// eslint-disable-next-line max-statements
export async function uploadPhotosIfNeeded (item: Item) {
  const data = structuredClone(item)
  const id = getItemId(data)
  if (!id.ok) return Result.error(id.error)

  for (const [index, photo] of data.photos.entries()) {
    if (!isUrl(photo)) continue
    let uuid = getAppWriteIdFromUrl(photo)
    if (uuid === undefined) {
      const result = await uploadImage(`${id.value}_photo-${index}`, photo)
      uuid = result.value
    }
    data.photos[index] = uuid
  }
  return Result.ok(data)
}

export function itemToAppWriteModel (item: Item) {
  const data = removeAppWriteFields(item)
  const result = safeParse(itemModelSchema, data)
  if (!result.success) return Result.error(result.issues.map(issue => issue.message).join(', '))
  return Result.ok(result.output)
}

// eslint-disable-next-line max-statements
export async function addItemRemotely (item: Item, currentState = state) {
  const data = await uploadPhotosIfNeeded(item)
  if (!data.ok) return Result.error(data.error)
  const id = getItemId(data.value)
  /* c8 ignore next */
  if (!id.ok) return Result.error(id.error)
  const { collectionId, databaseId } = currentState.credentials
  const payload = itemToAppWriteModel(data.value)
  if (!payload.ok) return payload
  const post = await Result.trySafe(database.createDocument<ItemModel>(databaseId, collectionId, id.value, payload.value))
  if (!post.ok) return post
  const parse = safeParse(itemSchema, post.value)
  if (!parse.success) return Result.error(parse.issues.map(issue => issue.message).join(', '))
  return Result.ok(parse.output)
}

export async function deleteItemRemotely (item: Item, currentState = state) {
  const { collectionId, databaseId } = currentState.credentials
  const result = await Result.trySafe(database.deleteDocument(databaseId, collectionId, item.$id))
  if (result.ok) for (const photo of item.photos) {
    if (isUrl(photo)) continue
    await deleteImageRemotely(photo)
  }
  return result
}

// eslint-disable-next-line max-statements
export async function updateItemRemotely (item: Item, currentState = state) {
  const data = await uploadPhotosIfNeeded(item)
  if (!data.ok) return Result.error(data.error)
  const { collectionId, databaseId } = currentState.credentials
  if (data.value.$id.length === 0) return Result.error(`item id is empty in ${JSON.stringify(data.value)}`)
  const payload = itemToAppWriteModel(data.value)
  if (!payload.ok) return payload
  const post = await Result.trySafe(database.updateDocument<ItemModel>(databaseId, collectionId, data.value.$id, payload.value))
  if (!post.ok) return post
  const parse = safeParse(itemSchema, post.value)
  if (!parse.success) return Result.error(parse.issues.map(issue => issue.message).join(', '))
  return Result.ok(parse.output)
}
