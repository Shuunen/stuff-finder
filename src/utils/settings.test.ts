import { expect, it } from 'vitest'
import { validateForm } from './forms.utils'
import { settingsForm } from './settings.utils'

it('validateForm A default settingsForm is not changed & not valid', () => {
  const { hasChanged, updatedForm } = validateForm(settingsForm)
  expect(updatedForm).toMatchSnapshot()
  expect(updatedForm.isTouched).toBe(false)
  expect(updatedForm.isValid).toBe(false)
  expect(updatedForm).toStrictEqual(settingsForm)
  expect(hasChanged).toBe(false)
})
