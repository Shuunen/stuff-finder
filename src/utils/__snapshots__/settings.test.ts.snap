// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validateForm A default settingsForm is not changed & not valid 1`] = `
{
  "errorMessage": "AppWrite database id is required",
  "fields": {
    "bucketId": {
      "columns": 1,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "AppWrite storage bucket id",
      "link": "https://appwrite.io/",
      "regex": /\\^\\[\\\\w-\\]\\{20,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "collectionId": {
      "columns": 1,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "AppWrite collection id",
      "link": "",
      "regex": /\\^\\[\\\\w-\\]\\{20,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
    "databaseId": {
      "columns": 1,
      "isRequired": true,
      "isValid": false,
      "isVisible": true,
      "label": "AppWrite database id",
      "link": "",
      "regex": /\\^\\[\\\\w-\\]\\{20,100\\}\\$/u,
      "type": "text",
      "unit": "",
      "value": "",
    },
  },
  "isTouched": false,
  "isValid": false,
}
`;
